Convolution derivatives in ATen

ezyang opened the pull request:
This PR introduces ATen implementation of convolution, which dispatches to THNN/CuDNN/nnpack based on input parameters. The general strategy is to compose this function out of the various forward-backward pairs of specific implementations, rather than write a monolithic function with backwards (which is what we did before because the boilerplate of doing it otherwise would have been very high.) The new API provides the following functions:

- `_convolution`, which is a fully generic, native convolution implementation that dispatches to various other convolution implementations depending on input characteristics. This is prefixed with an underscore because it explicitly takes `benchmark`, `deterministic` and `cudnn_enabled` which are implementation details for CuDNN. The intent is to eventually provide a `convolution` that reads these parameters out of the context using #4104.
- `_convolution_nogroup` is a convolution implementation for non-CuDNN algorithms which don't support group convolution natively.
- `_convolution_double_backward` is the generic double-backwards implementation for convolution.

This depends on #4096 (commits included).

In more detail:

- Most functionality from `torch/csrc/autograd/functions/convolution.cpp` has been moved into `aten/src/ATen/native/Convolution.cpp`
- We continue to make use of `ConvParams`, but we now construct the parameters upon entry to a function from the function signature (which does not use `ConvParams`; having `convolution` take `ConvParams` directly would require teaching the code generator how to accept these as parameters, complicating ATen's API model) and destruct them when making subprocedure calls.
- I introduce a new idiom, `input_r`, which represents a `const Tensor&` reference, which will subsequently be assigned to a local `Tensor input`. This is helpful because a lot of the existing algorithms relied on being able to assign to locals, which is not permitted with a const reference.
- The native argument parser now supports `std::array<bool,2>` inputs (NB: there MUST NOT be a space; this is the same hack as is applied to derivatives.yaml)
- Native parser now supports `Tensor?` arguments, which indicates a nullable tensor. Previously this function was only used by NN methods.
- Documentation updates on THNN library
- I added an extra `fgradInput` argument to `VolumetricConvolutionMM_updateOutput` and `VolumetricConvolutionMM_accGradParameters` so that its buffer list lines up with the backward argument list. This makes it possible to write derivative for `conv2d` which previously was not supported (commented out in `derivatives.yaml`)
- Extra `double_backward` declarations for all convolution backwards functions was added.
- Python argument parsing and `make_variable` augmented to handle undefined.

TODO:
- I bet NNPack doesn't work correctly

colesbury commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -70,7 +70,7 @@ FunctionParameter::FunctionParameter(const std::string& fmt, bool keyword_only)
   switch (type_) {
     case ParameterType::TENSOR: {
-      return THPVariable_Check(obj);
+      return obj == Py_None || THPVariable_Check(obj);
```
I don't think `None` checks should be part of this. They're handled for all types in the parsing.

colesbury commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -127,6 +127,7 @@ struct FunctionParameter {
 
 inline at::Tensor PythonArgs::tensor(int i) {
   if (!args[i]) return at::Tensor();
+  if (args[i] == Py_None) return at::Tensor();
```
`None` is transformed to `nullptr` in the parsing, so you shouldn't have this case.

ezyang replied:
The root cause was that I didn't set `allow_none = True`. In this situation, a `Py_None` will get propagated here.

colesbury commented on torch/csrc/utils/tuple_parser.cpp:
```diff
@@ -63,6 +63,22 @@ auto TupleParser::parse(std::vector<int>& x, const std::string& param_name) -> v
   }
 }
 
+auto TupleParser::parse(std::vector<int64_t>& x, const std::string& param_name) -> void {
```
Why is there a new `TupleParser` function. It looks like you are only removing uses of `TupleParser`?

colesbury commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -13,6 +13,7 @@ namespace torch { namespace autograd {
 
 Variable make_variable(at::Tensor data, std::shared_ptr<Function> grad_fn) {
   TORCH_ASSERT(grad_fn);
+  if (!data.defined()) return Variable();
```
Are these necessary for the convolution change?

ezyang replied:
Seems this is only needed if you return undefined from a function, which happened briefly when I was experimenting with having buffers be return values. The current patch doesn't need it, so I'll comment this out.

colesbury reviewed (commented):
looks pretty good. a few comments below

ezyang commented:
None argument parsing is now handled correctly.

Note that this is now a commit stack, so you have to look at Convolution derivatives in ATen  and subsequent commits.

apaszke commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -105,3 +106,14 @@
+- func: _convolution(Tensor input, Tensor weight, Tensor? bias, IntList stride, IntList padding, IntList dilation, bool transposed, IntList output_padding, int64_t groups, bool benchmark, bool deterministic, bool cudnn_enabled) -> Tensor
+  variants: function
+
+# NB: We MUST call the input self, otherwise codegen will attempt to
```
I think our current invariant is that all grads are always defined

ezyang replied:
Definitely not true today, but it's something that we are interested in fixing, it seems...

ezyang commented:
@pytorchbot retest this please

ezyang commented:
@pytorchbot retest this please

ezyang commented:
@pytorchbot retest this please

ezyang commented:
@pytorchbot retest this please

ezyang committed:
Convolution derivatives in ATen

This PR introduces ATen implementation of convolution, which dispatches to
THNN/CuDNN/nnpack based on input parameters. The general strategy is to compose
this function out of the various forward-backward pairs of specific
implementations, rather than write a monolithic function with backwards (which
is what we did before because the boilerplate of doing it otherwise would have
been very high.) The new API provides the following functions:

  - _convolution, which is a fully generic, native convolution implementation
    that dispatches to various other convolution implementations depending on
    input characteristics. This is prefixed with an underscore because it
    explicitly takes benchmark, deterministic and cudnn_enabled which are
    implementation details for CuDNN. The intent is to eventually provide a
    convolution that reads these parameters out of the context using #4104.
  - _convolution_nogroup is a convolution implementation for non-CuDNN
    algorithms which don't support group convolution natively.
  - _convolution_double_backward is the generic double-backwards implementation
    for convolution.

In more detail:

- Most functionality from torch/csrc/autograd/functions/convolution.cpp has been
  moved into aten/src/ATen/native/Convolution.cpp
- We continue to make use of ConvParams, but we now construct the parameters
  upon entry to a function from the function signature (which does not use
  ConvParams; having convolution take ConvParams directly would require teaching
  the code generator how to accept these as parameters, complicating ATen's API
  model) and destruct them when making subprocedure calls.
- I introduce a new idiom, input_r, which represents a const Tensor& reference,
  which will subsequently be assigned to a local Tensor input. This is helpful
  because a lot of the existing algorithms relied on being able to assign to
  locals, which is not permitted with a const reference.
- The native argument parser now supports std::array<bool,2> inputs (NB: there
  MUST NOT be a space; this is the same hack as is applied to derivatives.yaml)
- Native parser now supports Tensor? arguments, which indicates a nullable
  tensor. Previously this function was only used by NN methods.
- Documentation updates on THNN library
- I added an extra fgradInput argument to VolumetricConvolutionMM_updateOutput
  and VolumetricConvolutionMM_accGradParameters so that its buffer list lines up
  with the backward argument list. This makes it possible to write derivative
  for conv3d which previously was not supported (commented out in
  derivatives.yaml)
- Extra double_backward declarations for all convolution backwards functions was
  added.
- Python argument parsing and make_variable augmented to handle undefined.

TODO:

- This breaks convolution export for ONNX
- Only accept None in Python argument parser if is_nullable
- I bet NNPack doesn't work correctly

bugfix

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ONNX convolution

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Add missing include

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

minorfix

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

I think this fixes the problem.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

VolumetricConvolution fixup.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Undo bad change

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

chekcpoint

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

OK cudnn working now

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

make bias optional

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

lint

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Revert "Initialize on use."

This reverts commit fedbe5859d04651d24e9c59b037b7c15e4ee2509.

NNPack business.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Shunting around.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

bugfixes

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
@pytorchbot retest this please

colesbury commented on aten/src/ATen/UndefinedTensor.h:
```diff
@@ -7,6 +7,7 @@ namespace at {
 struct AT_API UndefinedTensor final : public TensorImpl {
 public:
   static inline UndefinedTensor * singleton() {
+    static UndefinedTensor _singleton;
```
This lazily initializes `_singleton`, which we don't want.

apaszke replied:
Why?

ezyang replied:
According to @gchanan, this is because a lazy static initializer means that the compiler must emit a conditional test to see if the variable is initialized or not.

Evidence: https://godbolt.org/g/a3NViC

apaszke replied:
Is that really such a big deal? It's not even an atomic instruction (1B `mov`), so I doubt it has _any_ impact on actual perf.

ezyang committed:
Build fix when nnpack is not available.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
more namespace fix

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
more fix

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Actually select nnpack

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

apaszke commented on aten/src/ATen/native/NNPACK.cpp:
```diff
@@ -0,0 +1,574 @@
+
+namespace at { namespace native {
+
+at::Tensor nnpack_spatial_convolution(
```
Eh we should figure out a better place for these files than a `native` directory. There's really nothing native in this code...

ezyang replied:
My understanding is that 'native' simply refers to any function whose binding code is implemented in ATen, as opposed to the legacy TH. This is why @zdevito has been so adamant about not exposing native to the public interface (we now have the abstract/concrete distinction for that). Does "native" have a meaning beyond "concrete"? If not, then perhaps what you are asking for is a rename. (I do think 'native' gives the wrong intuition).

ezyang committed:
lint

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
This is ready to land.

ezyang committed:
Delete obsolete NNPACK code.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

pietern commented:
@bddppq Looks like onnxbot found a failure here. Ideally the failure shows up here as one of the status checks, otherwise this signal is easily lost.

ezyang commented:
Yeah, this PR is known to wobble the ONNX expect files; I have a companion patch (not uploaded yet) for this

ezyang commented:
There appears to be a bug in the NNPACK codepath, not triggered by CI. Investigating.

ezyang committed:
Fix major NNPACK braino

Signed-off-by: Edward Z. Yang <ezyang@fb.com>
